File: src/portfolio_manager/repositories/stock_repository.py
"""Stock repository for database operations."""

from uuid import UUID, uuid4

from portfolio_manager.core.time import now_kst
from portfolio_manager.models import Stock
from portfolio_manager.services.database import StockModel


class StockRepository:
    """Repository for Stock database operations."""

    def create(self, ticker: str, group_id: UUID, name: str = "") -> Stock:
        """Create a new stock."""
        now = now_kst()
        row = StockModel.create(
            id=uuid4(),
            ticker=ticker,
            name=name,
            group=group_id,
            created_at=now,
            updated_at=now,
        )
        return self._to_domain(row)

    def list_by_group(self, group_id: UUID) -> list[Stock]:
        """List all stocks for a specific group."""
        return [
            self._to_domain(row)
            for row in StockModel.select().where(StockModel.group == group_id)
        ]

    def list_all(self) -> list[Stock]:
        """List all stocks."""
        return [self._to_domain(row) for row in StockModel.select()]

    def delete(self, stock_id: UUID) -> None:
        """Delete a stock by ID."""
        StockModel.delete().where(StockModel.id == stock_id).execute()

    def update(self, stock_id: UUID, ticker: str) -> Stock:
        """Update a stock ticker by ID."""
        now = now_kst()
        StockModel.update(ticker=ticker, updated_at=now).where(
            StockModel.id == stock_id
        ).execute()
        row = StockModel.get_by_id(stock_id)
        return self._to_domain(row)

    def get_by_id(self, stock_id: UUID) -> Stock | None:
        """Get a stock by ID."""
        row = StockModel.get_or_none(StockModel.id == stock_id)
        return self._to_domain(row) if row else None

    def get_by_ticker(self, ticker: str) -> Stock | None:
        """Get a stock by ticker."""
        row = StockModel.get_or_none(StockModel.ticker == ticker)
        return self._to_domain(row) if row else None

    def update_group(self, stock_id: UUID, group_id: UUID) -> Stock:
        """Update a stock's group by ID."""
        now = now_kst()
        StockModel.update(group=group_id, updated_at=now).where(
            StockModel.id == stock_id
        ).execute()
        row = StockModel.get_by_id(stock_id)
        return self._to_domain(row)

    def update_exchange(self, stock_id: UUID, exchange: str) -> Stock:
        """Update a stock's preferred exchange by ID."""
        now = now_kst()
        StockModel.update(exchange=exchange, updated_at=now).where(
            StockModel.id == stock_id
        ).execute()
        row = StockModel.get_by_id(stock_id)
        return self._to_domain(row)

    def update_name(self, stock_id: UUID, name: str) -> Stock:
        """Update a stock's display name by ID."""
        now = now_kst()
        StockModel.update(name=name, updated_at=now).where(
            StockModel.id == stock_id
        ).execute()
        row = StockModel.get_by_id(stock_id)
        return self._to_domain(row)

    @staticmethod
    def _to_domain(row: StockModel) -> Stock:
        return Stock(
            id=row.id,
            ticker=row.ticker,
            group_id=row.group_id,
            created_at=row.created_at,
            updated_at=row.updated_at,
            exchange=row.exchange,
            name=row.name,
        )

File: internal/repositories/stock_repository.go
package repositories

import (
	"context"
	"database/sql"
	"errors"

	"github.com/kadragon/portfolio-manager/internal/db/sqlc"
	"github.com/kadragon/portfolio-manager/internal/ktime"
	"github.com/kadragon/portfolio-manager/internal/models"
	"github.com/kadragon/portfolio-manager/internal/uuidx"
)

// StockRepository provides stock CRUD operations.
type StockRepository struct {
	q *sqlc.Queries
}

// NewStockRepository builds a StockRepository over the given queries handle.
func NewStockRepository(q *sqlc.Queries) *StockRepository {
	return &StockRepository{q: q}
}

// Create inserts a new stock with an empty name and no exchange.
func (r *StockRepository) Create(ctx context.Context, ticker string, groupID uuidx.UUID) (models.Stock, error) {
	now := ktime.Now()
	row, err := r.q.CreateStock(ctx, sqlc.CreateStockParams{
		ID:        uuidx.New(),
		Ticker:    ticker,
		GroupID:   groupID,
		Exchange:  sql.NullString{Valid: false},
		CreatedAt: now,
		UpdatedAt: now,
		Name:      "",
	})
	if err != nil {
		return models.Stock{}, err
	}
	return toStock(row), nil
}

// ListByGroup returns all stocks for the given group in insertion order.
func (r *StockRepository) ListByGroup(ctx context.Context, groupID uuidx.UUID) ([]models.Stock, error) {
	rows, err := r.q.ListStocksByGroup(ctx, groupID)
	if err != nil {
		return nil, err
	}
	out := make([]models.Stock, 0, len(rows))
	for _, row := range rows {
		out = append(out, toStock(row))
	}
	return out, nil
}

// ListAll returns every stock in insertion order.
func (r *StockRepository) ListAll(ctx context.Context) ([]models.Stock, error) {
	rows, err := r.q.ListAllStocks(ctx)
	if err != nil {
		return nil, err
	}
	out := make([]models.Stock, 0, len(rows))
	for _, row := range rows {
		out = append(out, toStock(row))
	}
	return out, nil
}

// GetByID returns the stock with the given id, or nil if absent.
func (r *StockRepository) GetByID(ctx context.Context, id uuidx.UUID) (*models.Stock, error) {
	row, err := r.q.GetStockByID(ctx, id)
	if errors.Is(err, sql.ErrNoRows) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	s := toStock(row)
	return &s, nil
}

// GetByTicker returns the stock with the given ticker, or nil if absent.
func (r *StockRepository) GetByTicker(ctx context.Context, ticker string) (*models.Stock, error) {
	row, err := r.q.GetStockByTicker(ctx, ticker)
	if errors.Is(err, sql.ErrNoRows) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	s := toStock(row)
	return &s, nil
}

// UpdateTicker changes the stock's ticker symbol.
func (r *StockRepository) UpdateTicker(ctx context.Context, id uuidx.UUID, ticker string) (models.Stock, error) {
	row, err := r.q.UpdateStockTicker(ctx, sqlc.UpdateStockTickerParams{
		Ticker:    ticker,
		UpdatedAt: ktime.Now(),
		ID:        id,
	})
	if err != nil {
		return models.Stock{}, err
	}
	return toStock(row), nil
}

// UpdateGroup moves the stock to a different group.
func (r *StockRepository) UpdateGroup(ctx context.Context, id uuidx.UUID, groupID uuidx.UUID) (models.Stock, error) {
	row, err := r.q.UpdateStockGroup(ctx, sqlc.UpdateStockGroupParams{
		GroupID:   groupID,
		UpdatedAt: ktime.Now(),
		ID:        id,
	})
	if err != nil {
		return models.Stock{}, err
	}
	return toStock(row), nil
}

// UpdateExchange sets the stock's exchange field.
func (r *StockRepository) UpdateExchange(ctx context.Context, id uuidx.UUID, exchange string) (models.Stock, error) {
	row, err := r.q.UpdateStockExchange(ctx, sqlc.UpdateStockExchangeParams{
		Exchange:  sql.NullString{String: exchange, Valid: true},
		UpdatedAt: ktime.Now(),
		ID:        id,
	})
	if err != nil {
		return models.Stock{}, err
	}
	return toStock(row), nil
}

// UpdateName sets the stock's display name.
func (r *StockRepository) UpdateName(ctx context.Context, id uuidx.UUID, name string) (models.Stock, error) {
	row, err := r.q.UpdateStockName(ctx, sqlc.UpdateStockNameParams{
		Name:      name,
		UpdatedAt: ktime.Now(),
		ID:        id,
	})
	if err != nil {
		return models.Stock{}, err
	}
	return toStock(row), nil
}

// Delete removes the stock with the given id.
func (r *StockRepository) Delete(ctx context.Context, id uuidx.UUID) error {
	return r.q.DeleteStock(ctx, id)
}

func toStock(row sqlc.Stock) models.Stock {
	var exchange *string
	if row.Exchange.Valid {
		exchange = &row.Exchange.String
	}
	return models.Stock{
		ID:        row.ID,
		Ticker:    row.Ticker,
		GroupID:   row.GroupID,
		Exchange:  exchange,
		Name:      row.Name,
		CreatedAt: row.CreatedAt.Time,
		UpdatedAt: row.UpdatedAt.Time,
	}
}
